relay-sampling/src/lib.rs

```diff
@@ -355,6 +365,125 @@ impl TimeRange {
     }
 }
 
+/// A struct representing the execution context of a decaying function.
+#[derive(Clone, Copy, Debug)]
+struct DecayingFunctionExecutor {
+    decaying_fn: DecayingFunction,
+    external_params: DecayingFunctionExternalParams,
+}
+
+impl DecayingFunctionExecutor {
+    fn execute(&self) -> f64 {
+        match self.decaying_fn {
+            DecayingFunction::Linear { from_sample_rate } => {
+                self.execute_linear_decay(from_sample_rate)
+            }
+            DecayingFunction::Constant => self.external_params.base_sample_rate,
+        }
+    }
+
+    /// Computes the linear decay function at a specific point in time.
+    ///
+    /// The linear decay works in the following way:
+    /// start = 100
+    /// end = 200
+    /// from = 0.7 (70% sample rate)
+    /// base_sample_rate = 0.5 (50% sample rate)
+    /// (from - base_sample_rate) = 0.2 (20% sample rate)
+    ///
+    /// If we have now = 170 it means we are at the 70% progress of the timerange, which means we
+    /// want to inversely reduce the sample rate in the interval (from - base_sample_rate).
+    /// Thus, in the 70% case, we would take only 30% of the (from - base_sample_rate) which
+    /// is equal to 0.06.
+    fn execute_linear_decay(&self, from_sample_rate: f64) -> f64 {
```

Comment:
nit: There are variables in this function that begin with `start_` and `end_`, so to clarify intent I would rename `from_sample_rate` to `start_sample_rate`, and create an alias `let end_sample_rate = self.external_params.base_sample_rate`.